Mathlib/Data/ListM/Basic.lean

```diff
@@ -361,6 +362,96 @@ This will run forever if the list is infinite. -/
 unsafe def fold (f : β → α → β) (init : β) (L : ListM m α) : m β :=
   L.foldM (fun b a => pure (f b a)) init
 
+/--
+Extract the prefix up a lazy list consisting of elements up to and including
+the first element satisfying a monadic predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+-/
+partial def getUpToFirstM (L : ListM m α) (p : α → m (ULift Bool)) : m (List α × ListM m α) := do
+  match ← L.uncons with
+  | none => return ([], nil)
+  | some (x, xs) => (if (← p x).down then
+      return ([x], xs)
+    else do
+      let (acc, R) ← getUpToFirstM xs p
+      return (x :: acc, R))
+
+/--
+Extract the prefix up a lazy list consisting of elements up to and including
+the first element satisfying a predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+-/
+def getUpToFirst (L : ListM m α) (p : α → Bool) : m (List α × ListM m α) :=
+  L.getUpToFirstM fun a => pure (.up (p a))
+
+/--
+Extract a maximal prefix of a lazy list consisting of elements
+satisfying a monadic predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+
+(Note that the first element *not* satisfying the predicate will be generated,
+and pushed back on to the remaining lazy list.)
+-/
+partial def splitWhileM (L : ListM m α) (p : α → m (ULift Bool)) :
+    m (List α × ListM m α) := do
+  match ← L.uncons with
+  | none => return ([], nil)
+  | some (x, xs) => (if (← p x).down then do
+      let (acc, R) ← splitWhileM xs p
+      return (x :: acc, R)
+    else
+      return ([], cons do pure (some x, xs)))
+
+/--
+Extract a maximal prefix of a lazy list consisting of elements
+satisfying a predicate.
+Return (in the monad) the prefix as a `List`, along with the remaining elements as a `ListM`.
+
+(Note that the first element *not* satisfying the predicate will be generated,
+and pushed back on to the remaining lazy list.)
+-/
+def splitWhile (L : ListM m α) (p : α → Bool) : m (List α × ListM m α) :=
+  L.splitWhileM fun a => pure (.up (p a))
+
+/--
+Splits a lazy list into contiguous sublists of elements with the same value under
+a monadic function.
+Return a lazy lists of pairs, consisting of a value under that function,
+and a maximal list of elements having that value.
+-/
+partial def runsM [DecidableEq β] (L : ListM m α) (f : α → m β) : ListM m (β × List α) :=
+  L.cases' nil fun a t => squash do
+    let b ← f a
+    let (l, t') ← t.splitWhileM (fun a => do return .up ((← f a) = b))
+    return cons do pure (some (b, a :: l), t'.runsM f)
+
+/--
+Splits a lazy list into contiguous sublists of elements with the same value under a function.
+Return a lazy lists of pairs, consisting of a value under that function,
+and a maximal list of elements having that value.
+-/
+def runs [DecidableEq β] (L : ListM m α) (f : α → β) : ListM m (β × List α) :=
```

Comment:
This looks a lot like python's "itertools.groupby"

Reply:
Oh, I do like `groupBy`!